File: examples/e-spring-example/src/main/java/dev/akif/espringexample/people/PeopleRepository.java
package dev.akif.espringexample.people;

import java.util.List;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import dev.akif.espringexample.errors.Errors;
import dev.akif.espringexample.crud.Repository;
import dev.akif.espringexample.people.dto.PersonDTO;
import dev.akif.espringexample.people.model.Person;
import e.java.E;
import e.java.EOr;

@Component
public class PeopleRepository implements Repository<Person, PersonDTO> {
    private final PeopleJPARepository jpa;

    @Autowired
    public PeopleRepository(PeopleJPARepository jpa) {
        this.jpa = jpa;
    }

    @Override public EOr<List<Person>> getAll() {
        return EOr.catching(
            () -> toList(jpa.findAll()),
            t  -> Errors.database.message("Cannot get all people!").cause(E.fromThrowable(t))
        );
    }

    @Override public EOr<Optional<Person>> getById(long id) {
        return EOr.catching(
            () -> jpa.findById(id),
            t  -> Errors.database.message("Cannot get person by id!").cause(E.fromThrowable(t)).data("id", id)
        );
    }

    @Override public EOr<Person> create(PersonDTO personDTO) {
        return EOr.catching(
            () -> jpa.save(personDTO.toPerson()),
            t  -> Errors.database
                        .message("Cannot create person!")
                        .cause(E.fromThrowable(t))
                        .data("name", personDTO.getName())
                        .data("age",  personDTO.getAge())
        );
    }

    @Override public EOr<Person> update(long id, PersonDTO personDTO) {
        return EOr.catching(
            () -> jpa.findById(id),
            t -> Errors.database
                       .message("Cannot find person to update!")
                       .cause(E.fromThrowable(t))
                       .data("id", id)
        ).flatMap(personOpt ->
            EOr.fromOptional(personOpt, () ->
                Errors.notFound
                      .message("Cannot find person to update!")
                      .data("id", id)
            )
        ).flatMap(person ->
            EOr.catching(
                () -> {
                    Person updated = personDTO.updated(person);
                    jpa.save(updated);
                    return updated;
                },
                t -> Errors.database
                           .message("Cannot update person!")
                           .cause(E.fromThrowable(t))
                           .data("name", personDTO.getName())
                           .data("age",  personDTO.getAge())
            )
        );
    }

    @Override public EOr<Person> delete(long id) {
        return EOr.catching(
            () -> jpa.findById(id),
            t -> Errors.database
                       .message("Cannot find person to delete!")
                       .cause(E.fromThrowable(t))
                       .data("id", id)
        ).flatMap(personOpt ->
            EOr.fromOptional(personOpt, () ->
                Errors.notFound
                      .message("Cannot find person to delete!")
                      .data("id", id)
            )
        ).flatMap(person ->
            EOr.catching(
                () -> {
                    jpa.delete(person);
                    return person;
                },
                t -> Errors.database
                           .message("Cannot delete person!")
                           .cause(E.fromThrowable(t))
                           .data("id", id)
            )
        );
    }
}

File: examples/e-http4s-example/src/main/scala/dev/akif/ehttp4sexample/people/PeopleRepository.scala
package dev.akif.ehttp4sexample.people
import e._
import cats.effect.IO
import dev.akif.ehttp4sexample.common.implicits._
import dev.akif.ehttp4sexample.common.{Errors, Repository}
import doobie.implicits._
import doobie.util.transactor.Transactor

class PeopleRepository(override val db: Transactor[IO]) extends Repository[IO, Person, CreatePerson, UpdatePerson] {
  override def getAll: IO[List[Person]] = {
    val operation =
      sql"""SELECT id, name, age FROM people ORDER BY name""".query[Person].to[List]

    operation.transact(db).handleErrorWith { t =>
      Errors.database
            .message("Cannot get people!")
            .cause(t.toE())
            .toIO
    }
  }

  override def get(id: Long): IO[Option[Person]] = {
    val operation =
      sql"""SELECT id, name, age FROM people WHERE id = $id""".query[Person].option

    operation.transact(db).handleErrorWith { t =>
      Errors.database
            .message("Cannot get person!")
            .data("id" -> id)
            .cause(t.toE())
            .toIO
    }
  }

  override def create(create: CreatePerson): IO[Person] = {
    val operation =
      sql"""INSERT INTO people(name, age) VALUES (${create.name}, ${create.age})""".update.withUniqueGeneratedKeys[Long]("id")

    operation.transact(db).map(create.toPerson).handleErrorWith { t =>
      Errors.database
            .message("Cannot create person!")
            .data("name" -> create.name)
            .data("age" -> create.age)
            .cause(t.toE())
            .toIO
    }
  }

  override def update(id: Long, update: UpdatePerson): IO[Person] = {
    val operation =
      for {
        personOpt  <- sql"""SELECT id, name, age FROM people WHERE id = $id""".query[Person].option
        person     <- handleOption(personOpt, Errors.notFound.message("Cannot find person to update!").data("id" -> id))
        updated     = update.updated(person)
        _          <- sql"""UPDATE people SET name = ${updated.name}, age = ${updated.age} WHERE id = $id""".update.run
      } yield {
        updated
      }

    operation.transact(db).handleErrorWith { t =>
        Errors.database
          .message("Cannot update person!")
          .data("name" -> update.name)
          .data("age" -> update.age)
          .cause(t.toE())
          .toIO
      }
  }

  override def delete(id: Long): IO[Person] = {
    val operation =
      for {
        personOpt  <- sql"""SELECT id, name, age FROM people WHERE id = $id""".query[Person].option
        person     <- handleOption(personOpt, Errors.notFound.message("Cannot find person to delete!").data("id" -> id))
        _          <- sql"""DELETE FROM people WHERE id = $id""".update.run
      } yield {
        person
      }

    operation.transact(db).handleErrorWith { t =>
      Errors.database
        .message("Cannot delete person!")
        .data("id" -> id)
        .cause(t.toE())
        .toIO
    }
  }
}
